[MHLO-Linalg] Lower mhlo.convolution to linalg when padding is non zero.

arnab-polymage opened the pull request:
Lower mhlo.convolution to linalg when padding is non zero.

joker-eph commented on tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc:
```diff
@@ -1729,6 +1761,27 @@ struct NormalConvOpConversion : public OpConversionPattern<mhlo::ConvOp> {
+          strides, dilations, PruneAttributeList(op));
+      rewriter.replaceOp(op, convOp.getOperation()->getResults());
+      return success();
+    }
```
Why the special case for rank 4 only?

arnab-oss replied:
Sure. We can extend this for rank 3 and rank 4 too

joker-eph commented on tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc:
```diff
@@ -1729,6 +1761,27 @@ struct NormalConvOpConversion : public OpConversionPattern<mhlo::ConvOp> {
     // input dilation for deconvolution cases.
     Attribute dilations = op.rhs_dilationAttr();
+
+    // Check if padding is zero.
```
I think you should improve the comment here, it should be a few lines to describe what's happening here, possibly explaining how the padding is handled on the input.

joker-eph commented:
If I read the patch correctly: we are currently lowering these test cases but incorrectly so? 
So this is a bug fix?

Can you elaborate in the pull-request description?

joker-eph commented on tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc:
```diff
@@ -1679,6 +1679,44 @@ struct PadOpConversion : public OpConversionPattern<mhlo::PadOp> {
+/// Extract element values from `attr` and store them in `arrayValues`.
+static void getValuesFromIntAttribute(DenseIntElementsAttr attr,
+                                      SmallVector<int64_t, 4>& arrayValues) {
+  for (auto val : attr.getValues<APInt>())
```
You should be able to do something like:

```suggestion
  for (auto val : attr.getValues<int64_t>())
```

arnab-oss replied:
I think this is a useful helper function, which may have future uses. It no way depends on the lowering which I'm implementing.

joker-eph reviewed (commented):
Sorry: I wrote some minor inline comment last time and didn't send them out...

LG overall.

arnab-oss commented on tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc:
```diff
@@ -1679,6 +1679,44 @@ struct PadOpConversion : public OpConversionPattern<mhlo::PadOp> {
+static Value applyPad(Location loc, Value input, ArrayRef<int64_t> pad,
+                      Attribute padAttr, OpBuilder& rewriter) {
+  // Input should be padded if necessary.
+  if (llvm::all_of(pad, [](int64_t p) { return p == 0; })) return input;
```
Actually this condition is redundant....I'm already checking this condition in the caller function

arnab-polymage committed:
Lower mhlo.convolution to linalg when padding is non zero.

Lower mhlo.convolution to linalg when padding is non zero into a
combination of linalg conv ops and linalg pad_tensor op. We apply
padding to the input first, and then pass the padded input while
creating linalg convolution ops.
